### PR TITLE
Stable diffusion with aitemplate example

### PR DESCRIPTION
**Pull request by maxscheel**

### CODE REVIEW

File: 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        setup_dockerfile_commands=[
+            "RUN apt-get update && apt-get install -y git python3-pip",
+            "RUN ln -s /usr/bin/python3 /usr/bin/python",
+            "WORKDIR /app",
+            "RUN git clone --recursive https://github.com/facebookincubator/AITemplate.git",
+            "WORKDIR /app/AITemplate/python",
+            "RUN git checkout b041e0e",
+            "RUN python setup.py bdist_wheel && pip install dist/aitemplate-*.whl",
+            "WORKDIR /app/AITemplate/examples/05_stable_diffusion",
+            "RUN sed -i src/pipeline_stable_diffusion_ait.py -e 's/unet.in_channels/unet.config.in_channels/g'",
```
Comment (luiscape) [inline 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py:160]: Could you add a comment on why this is required?

Comment (maxscheel) [reply 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py:163]: I added some comments in [3c12ed8](https://github.com/modal-labs/modal-examples/pull/308/commits/3c12ed89abb717da88e4439ec0c79ee5a3e76408)

File: 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+#
+# This example shows the Stable Diffusion 1.5 compiled with [AITemplate](https://github.com/facebookincubator/AITemplate) to run faster on Modal.
+# There is also a [Stable Diffusion CLI example](/docs/guide/ex/stable_diffusion_cli).
+# 
+# #### Upsides
+#  - Image generation improves over the CLI example to about 550ms per image generated (A10G, 10 steps, 512x512, png).
+# 
+# #### Downsides
+#  - Width and height as well as batch size must be configured prior to compilation which takes about 15 minutes.
+#  - In this example the compilation is done at docker image creation.
```
Comment (thundergolfer) [inline 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py:11]:
```suggestion
#  - In this example the compilation is done during image creation.
```

Our image building does not use Docker.

File: 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
+# # Stable Diffusion (AITemplate Edition)
+#
+# This example shows the Stable Diffusion 1.5 compiled with [AITemplate](https://github.com/facebookincubator/AITemplate) to run faster on Modal.
+# There is also a [Stable Diffusion CLI example](/docs/guide/ex/stable_diffusion_cli).
+# 
+# #### Upsides
+#  - Image generation improves over the CLI example to about 550ms per image generated (A10G, 10 steps, 512x512, png).
```
Comment (thundergolfer) [inline 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py:7]: Might be good to add the CLI example's perf number.

File: 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+gpu_type = "A10G"
+model_id = "runwayml/stable-diffusion-v1-5"
+width = 512
+height = 512
+batch_size = 1
+
+# ## Set the paths to the AITemplate example. This is needed to import the compiler and pipeline.
+
+def set_paths():
+    ait_sd_example_path = "/app/AITemplate/examples/05_stable_diffusion"
```
Comment (thundergolfer) [inline 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py:35]: I see this is the same path ref'd at https://github.com/modal-labs/modal-examples/pull/308/files#diff-40fda8df91885f0f4c10fdd86e8c1f827316b8da25b0b944f7033f180deb6cb3R161

Could break it out into a `workdir` global config variable?

File: 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+class InferenceRequest(BaseModel):
+    prompt: str = "photo of a cat in the snow, blue eyes, highly detailed, 8k, 200mm canon lens, shallow depth of field"
+    num_inference_steps: int = 10
+    guidance_scale: float = 7.5
+    negative_prompt: str = "deformed, extra legs, no tail"
+    format: str = "webp"
+
+stub = modal.Stub()
```
Comment (thundergolfer) [inline 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py:208]:
```suggestion
stub = modal.Stub(name="example-stable-diffusion-aitemplate")
```

File: 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+# This example shows the Stable Diffusion 1.5 compiled with [AITemplate](https://github.com/facebookincubator/AITemplate) to run faster on Modal.
+# There is also a [Stable Diffusion CLI example](/docs/guide/ex/stable_diffusion_cli).
+# 
+# #### Upsides
+#  - Image generation improves over the CLI example to about 550ms per image generated (A10G, 10 steps, 512x512, png).
+# 
+# #### Downsides
+#  - Width and height as well as batch size must be configured prior to compilation which takes about 15 minutes.
+#  - In this example the compilation is done at docker image creation.
+#  - Cold start time are also increased to upto ~30s.
```
Comment (thundergolfer) [inline 06_gpu_and_ml/stable_diffusion_aitemplate/stable_diffusion_aitemplate.py:12]:
```suggestion
#  - Cold start time are also increased to up-to ~30s from ~10s.
```

**Review by thundergolfer — COMMENTED** [review]
Thanks for the contribution 💯 

Main question I have is whether this file can be placed in `stable_diffusion/`. Trying to group examples by model type, and not have a subdirectory per-example.

**Comment by luiscape** [comment]
I've made a few modifications in [this PR](https://github.com/modal-labs/modal-examples/compare/main...luiscape:stable-diffusion-aitemplate?expand=1). Closing in favor of that one.

**Pull request closed without merging**